Migrate creature positions to absolute tripoints

eltank opened the pull request:
#### Summary
Infrastructure "Migrate creature positions to absolute tripoints"

#### Purpose of change

Typed tripoints are less error-prone compared to raw tripoints since their meaning can be inferred from the type and their proper usage enforced by the compiler.
Absolute coordinates (relative to a global origin point) are less error-prone compared to relative coordinates due to not requiring careful adjustments when the point of origin (typically the reality bubble) moves around.
This PR changes the internal and external representation of creature positions (for monsters, npcs, and avatar) to `tripoint_abs_ms`, a typed absolute coordinate system.
It also happens to fix a bug with horde behavior after entering the reality bubble (see below).

**Warning: Saves created after this PR cannot be loaded by earlier versions of the game!**

#### Describe the solution

- Introduce a `tripoint_abs_ms` `location` field to replace `tripoint` `position` in the `Creature` class.
- Introduce `move_to` as an alternative to `setpos`, the former taking in `tripoint_abs_ms` for the destination. Make `setpos` effectively delegate to the new function.
- Refactor side-effects of moving a creature into a protected `on_move` function invoked by both `setpos` and `move_to`. This is overridden by subclasses.
- Remove `submap_coords` from the `npc` class.
- Remove "position shifting" from `npc` and `monster` (used to be done when the reality bubble shifts by 1 submap). The shift functions are still present and active but operate only on fields that aren't serialized [to be migrated later]
- Refactor `creature_tracker` to use `tripoint_abs_ms` as the key for its location map. Also remove the forced cache rebuild that used to happen on map shift.
- Add function overloads taking in `tripoint_abs_ms` to the `map` class. Also added `map::getglobal()` which returns `tripoint_abs_ms`.
- Migrated other location fields from `monster`: `goal`, `wander_pos`, `patrol_route`. 
- Removed the serialization for `monster::path` and added `path.clear()` during map shifting. This is necessary because `path` is map-relative and will end up pointing to wrong positions after a shift. I copied the shift invalidation fix from `npc`. `path` doesn't need to be serialized because it's essentially a cache for pathfinding to `goal`, so removing it reduces save file bloat.
- Migrated other location fields from `npc`: `last_player_seen_pos`, `guard_pos`, `pulp_location`, `chair_pos`, `wander_pos`

#### Describe alternatives you've considered

Doing this for vehicles too. I think it's best to leave it for another PR.

While working on this I noticed that inactive non-horde monsters are stored in the overmap save files, taking up 95% of the space there (which means they're also always in memory for currently loaded overmaps). We can improve performance by storing them in the submap save files (loaded on demand). Another PR.

While looking at the save files I also noticed that we output a lot of unnecessary fields such as `"speed_bonus":0,"dodge_bonus":0.000000,"block_bonus":0` or `"mounted_player_id":-1` or `"damage_bandaged":0,"damage_disinfected":0`. These values which are essentially unset don't need to be stored in the save file. Also since when do monsters bandage and disinfect themselves? Ah well, stuff to clean up some other time.

I noticed that when zombies re-enter the reality bubble their goal is cleared. Now that the goal is in absolute coordinates and won't get corrupted by re-spawning the zombie from a different map area we can probably stop clearing the goal.

#### Testing

In a version of the game without my change I created 3 instances of "guarded chicken" (i.e. 1 fenced-in monster + 1 npc on guard duty):
![image](https://user-images.githubusercontent.com/8000047/132083223-a562f54d-f249-4b5e-86ba-ebe2fdba6c45.png)

I put 2 in the reality bubble, one in the same submap as the player and one several submaps away to the east. The 3rd one I put far south outside of the reality bubble.
After loading the save with a new binary (which would trigger migration) I walked around to each location to make sure the NPC and chicken were still where they were supposed to be. This caught some bugs during the early phases of the PR (disappearing or teleporting creatures) but now they are all good. For the chicken outside the reality bubble I tried approaching from the north (reversing the path I took to get away from it) as well as circling around and approaching from the east. This makes it enter the reality bubble at different map-local coordinates.
[Test.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/7110664/Test.zip)

To test critters fleeing and leaving the reality bubble, I spawned a deer and watched it run away until I couldn't see it any more. Then I teleported after it to make sure it was not gone.

Spawned a few zombies, let them see me, then went invisible, no scent, no sound and walked away a bit. They clustered where they'd seen me then started randomly dispersing. Made a sound and they chased that for a bit. Turned on scent and they followed the scent trail as usual.

Spawned a grappler and let it grab me to make sure grabbing still works.
Spawned a garghoul to make sure leaping still works.
Spawned a creeper hub and watched vines grow. Killed the hub and vines died.
Placed an autodoc on the map, spawned anesthesia and a Power CBM, installed the CBM. Spawned a "bad nurse" and let it grab me. It knocked me out and tried to remove the CBM on the spot but failed. I was expecting to be dragged to the autodoc couch, but that didn't happen. I tried with an older binary and the same thing happened. It looks like the lack of dragging is an existing bug or limitation.

I tested horde behavior. To my mild surprise, this PR fixes a major bug. Here's a test world in which I made a lot of noise near a city using a 50-cal. The debug "show horde" option shows several hordes with lots of zombies interested in my location. Yet without this PR what happens is that as soon as any zombies enter the reality bubble (which happens visibly BTW, not off-screen as it's supposed to) they veer off to the west. After 2 hours of waiting, none of them reach my position because they're going to some incorrectly-computed destination.
[HordeTest.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/7110881/HordeTest.zip)

![image](https://user-images.githubusercontent.com/8000047/132114961-a5602b99-3b46-41c8-a21f-1cd885204f54.png)

Loading the same world after this PR yields a much more reasonable outcome after waiting (note, they can't see me due to debug mutations):

![image](https://user-images.githubusercontent.com/8000047/132115054-8401f49f-25c9-4bdf-9cb5-c33d40078f6c.png)

This is because of the changes to `map::spawn_monsters_submap_group` where I introduced a new argument to pass down absolute submap coordinates from the caller. I believe the old code computes a destination relative to the wrong overmap (which illustrates nicely the value of absolute strongly-typed coordinates).

Not sure what else to test. Suggestions welcome.

#### Additional context

There are over 2300 references to `Creature::pos()` (which should migrate to `get_location()`) so that's not going away any time soon. There are "only" 125 calls to `setpos()` so getting rid of that is more feasible, but will still take multiple PRs.

Removal highlights:
```
    // The order in which zombies are shifted may cause zombies to briefly exist on
    // the same square. This messes up the mon_at cache, so we need to rebuild it.

    // The global absolute position (in map squares) of the npc is *always*
    // "submap_coords.x * SEEX + posx() % SEEX" (analog for y).
    // The main map assumes that pos is in its own (local to the main map)
    // coordinate system. We have to change pos to match that assumption

     * For active NPCs pos would be in the valid range required by
     * the map. But pos, map, and om can be changed without the NPC
     * actual moving as long as the position stays the same:
     * pos() += SEEX; submap_coords.x -= 1;

    // The absolute position in map squares, (x,y) is already global, but it's a
    // submap coordinate, so translate it and add the exact monster position on
    // the submap. modulo because the zombies position might be negative, as it
    // is stored *after* it has gone out of bounds during shifting.
```

eltank commented:
I've decided to migrate the rest of the monster and npc fields after all, so this is going to grow a bit more. The reason for that is I think splitting the migration has a higher chance of introducing bugs due to inconsistencies.

eltank commented:
Found a bug while chasing deer. When a monster reaches a "goal" point after fleeing it becomes stuck in that spot. It can randomly move away but it will immediately go back to it if nothing else forces it to change "goal". I think I know where the problem is.

eltank commented:
@jbytheway Please take a look if you have time.

jbytheway commented:
I'll try to look tomorrow.

actual-nh commented:
@eltank: Just to ping you on this - you have conflicts, probably due to things moving out of character.cpp.

eltank commented:
> @eltank: Just to ping you on this - you have conflicts, probably due to things moving out of character.cpp.

Rebased and fixed. Got lucky this time, it wasn't a real conflict, just something adjacent to my edits. But the longer this PR sits, the more conflicts it will accumulate. Hopefully it can go in soon.

jbytheway commented on src/faction_camp.cpp:
```diff
@@ -627,7 +627,7 @@ void talk_function::basecamp_mission( npc &p )
         }
         tripoint src_loc;
-        const tripoint abspos = p.global_square_location().raw();
+        const tripoint abspos = p.get_location().raw();
```
```suggestion
        // TODO: fix point types
        const tripoint abspos = p.get_location().raw();
```

eltank replied:
done

jbytheway commented on src/map.cpp:
```diff
@@ -7455,26 +7454,25 @@ void map::spawn_monsters_submap_group( const tripoint &gp, mongroup &group, bool
-                const tripoint rand_dest = horde_target +
-                                           point( rng( 0, SEEX ), rng( 0, SEEY ) );
-                const int turns = rl_dist( p, rand_dest ) + group.interest;
+                const tripoint_abs_ms rand_dest = project_to<coords::ms>( horde_target ) + point( rng( 0, SEEX ),
+                                                  rng( 0, SEEY ) );
```
This line has grown to be rather ugly.  Consider e.g.
```suggestion
                point_sm_ms pos_in_sm( rng( 0, SEEX ), rng( 0, SEEY ) );
                const tripoint_abs_ms rand_dest = project_combine( horde_target, pos_in_sm );
```

eltank replied:
Done (and you forgot const :)

eltank replied:
Hmm, it also needs to be `point_rel_ms`, `point_sm_ms` doesn't work.

jbytheway replied:
If you're just adding them then you'd need `point_rel_ms`m but I thought `point_sm_ms` would work with `project_combine`?

jbytheway commented on src/map.cpp:
```diff
@@ -8089,11 +8091,21 @@ tripoint map::getabs( const tripoint &p ) const
 
+tripoint map::getlocal( const tripoint_abs_ms &p ) const
+{
+    return getlocal( p.raw() );
```
```suggestion
    // TODO: fix point types
    return getlocal( p.raw() );
```

eltank replied:
Done

jbytheway commented on src/monster.cpp:
```diff
@@ -3070,11 +3092,14 @@ void monster::hear_sound( const tripoint &source, const int vol, const int dist,
-                   wander_turns );
+        // TODO: make the destination scale with the sound and handle
+        // the case when (x,y) is the same by picking a random direction
+        tripoint_abs_ms away = get_location() - target + get_location();
```
Took me a while to figure out why this strange computation.  Might warrant a comment.
```suggestion
        // Move towards a point on the opposite side of us from the target
        tripoint_abs_ms away = get_location() + ( get_location() - target );
```

eltank replied:
Done.

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -190,6 +190,17 @@ static void deserialize( weak_ptr_fast<monster> &obj, JsonIn &jsin )
+static tripoint read_legacy_creature_pos( const JsonObject &data )
+{
+    tripoint pos;
+    if( !data.read( "posx", pos.x ) ) {
+        debugmsg( "Bad Creature JSON: neither `location` nor `posx` found" );
+    }
+    data.read( "posy", pos.y );
+    data.read( "posz", pos.z );
```
```suggestion
    if( !data.read( "posx", pos.x ) || !data.read( "posy", pos.y ) || !data.read( "posz", pos.z ) ) {
        debugmsg( R"(Bad Creature JSON: neither "location" nor "posx", "posy", "posz" found)" );
    }
```
Backticks don't really make sense in this context, but I understand why you chose them :).

eltank replied:
Thanks for the tip about R"()".
Done.

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -1324,6 +1335,11 @@ void avatar::load( const JsonObject &data )
 {
     Character::load( data );
 
+    // Remove after 0.G
```
```suggestion
    // TEMPORARY until 0.G
```
All-caps `TEMPORARY` is what we agreed at some point as the thing to be able to grep for on new releases; I realise that other comments here don't match that style.

eltank replied:
done

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -1743,6 +1759,39 @@ void npc::load( const JsonObject &data )
 {
     Character::load( data );
 
+    // Remove if branch after 0.G (keep else branch)
```
```suggestion
    // TEMPORARY Remove if branch after 0.G (keep else branch)
```

eltank replied:
done

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -2104,33 +2102,74 @@ void inventory::json_load_items( JsonIn &jsin )
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 ///// monster.h
 
+// Remove after 0.G
```
```suggestion
// TEMPORARY until 0.G
```

eltank replied:
done

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -2104,33 +2102,74 @@ void inventory::json_load_items( JsonIn &jsin )
     load( data );
 }
 
+// Remove after 0.G
```
```suggestion
// TEMPORARY until 0.G
```

eltank replied:
done

jbytheway commented on src/savegame_json.cpp:
```diff
@@ -2104,33 +2102,74 @@ void inventory::json_load_items( JsonIn &jsin )
 {
     Creature::load( data );
 
+    // Remove after 0.G
```
```suggestion
    // TEMPORARY until 0.G
```

eltank replied:
done

jbytheway reviewed (commented):
This is awesome work.  I didn't see any notable flaws.

Some minor comments below.

In particular, I've been trying to add `// TODO: fix point types` at places where the "old" and "new" point types interact, especially for situations where there may not be a compile error when the types change.  This is partly to make it easy to find places to continue the point migration effort.

eltank commented on src/creature_tracker.cpp:
```diff
@@ -69,16 +70,15 @@ bool creature_tracker::add( const shared_ptr_fast<monster> &critter_ptr )
         } else {
-            debugmsg( "there's already a monster at %d,%d,%d", critter.pos().x, critter.pos().y,
-                      critter.pos().z );
+            debugmsg( "there's already a monster at %s", critter.get_location().to_string_writable() );
```
Well, in a past PR I used to_string() and clang-tidy complained so I've been using to_string_writable() since then.

jbytheway replied:
`to_string_writable` is needed for `add_msg_debug` (and friends), but not for `debugmsg` (I know, it's confusing).  But on reflection it probably makes sense to use it for `debugmsg` too, so no need to change your code here.

eltank commented on src/creature_tracker.cpp:
```diff
@@ -296,6 +294,12 @@ void creature_tracker::remove_dead()
 template<typename T>
 T *creature_tracker::creature_at( const tripoint &p, bool allow_hallucination )
+{
+    return creature_at<T>( tripoint_abs_ms( get_map().getabs( p ) ), allow_hallucination );
```
Ah, this should just use `map::getglobal()`. I added that function later but didn't change all of my earlier code to use it. [done]
Of course, this creature_at overload taking in a raw tripoint should be removed when possible -- which will be after migrating the map interface to take absolute tripoints.

eltank commented on src/creature_tracker.cpp:
```diff
@@ -330,17 +334,36 @@ T *creature_tracker::creature_at( const tripoint &p, bool allow_hallucination )
 template<typename T>
 const T *creature_tracker::creature_at( const tripoint &p, bool allow_hallucination ) const
+{
+    return creature_at<T>( tripoint_abs_ms( get_map().getabs( p ) ), allow_hallucination );
```
changed to getglobal()

eltank commented on src/creature.cpp:
```diff
@@ -123,16 +124,37 @@ Creature &Creature::operator=( Creature && ) noexcept = default;
 
+tripoint Creature::pos() const
+{
+    return get_map().getlocal( location.raw() );
```
dropped the "raw()"

eltank commented on src/creature.cpp:
```diff
@@ -123,16 +124,37 @@ Creature &Creature::operator=( Creature && ) noexcept = default;
 void Creature::set_pos_only( const tripoint &p )
 {
-    position = p;
+    location = tripoint_abs_ms( get_map().getabs( p ) );
```
switched to getglobal()

eltank commented:
npc_test started failing, looking into it

jbytheway reviewed (approved):
LGTM
